sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTables.scala

```diff
@@ -0,0 +1,190 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.plans.logical.{AlterTable, DeleteFromTable, DescribeTable, LogicalPlan, SubqueryAlias}
+import org.apache.spark.sql.catalyst.plans.logical.sql.{AlterTableAddColumnsStatement, AlterTableAlterColumnStatement, AlterTableDropColumnsStatement, AlterTableRenameColumnStatement, AlterTableSetLocationStatement, AlterTableSetPropertiesStatement, AlterTableUnsetPropertiesStatement, AlterViewSetPropertiesStatement, AlterViewUnsetPropertiesStatement, DeleteFromStatement, DescribeTableStatement, QualifiedColType}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.{CatalogManager, LookupCatalog, TableChange}
+import org.apache.spark.sql.execution.command.{AlterTableAddColumnsCommand, AlterTableSetLocationCommand, AlterTableSetPropertiesCommand, AlterTableUnsetPropertiesCommand, DescribeTableCommand}
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.types.{HIVE_TYPE_STRING, HiveStringType, MetadataBuilder, StructField}
+
+/**
+ * Resolves tables from the multi-part identifiers in DDL/DML commands.
+ *
+ * For each SQL statement, this rule has 2 different code paths for v1 and v2 tables.
+ */
+class ResolveTables(val catalogManager: CatalogManager)
+  extends Rule[LogicalPlan] with LookupCatalog {
+  import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+    case AlterTableAddColumnsStatement(
+         CatalogAndTable(catalog, tblName, Left(v1Table)), cols) =>
+      cols.foreach(c => assertTopLeveColumn(c.name, "AlterTableAddColumnsCommand"))
+      AlterTableAddColumnsCommand(tblName.toV1Identifier, cols.map(convertToStructField))
+
+    case AlterTableAddColumnsStatement(
+         CatalogAndTable(catalog, tblName, Right(table)), cols) =>
+      val changes = cols.map { col =>
+        TableChange.addColumn(col.name.toArray, col.dataType, true, col.comment.orNull)
+      }
+      AlterTable(catalog, tblName.toIdentifier, table, changes)
+
+    // The v1 `AlterTableAddColumnsCommand` will check temp view and provide better error message.
+    // Here we convert the statement to the v1 command to get the better error message.
+    // TODO: apply the temp view check for all ALTER TABLE statements.
+    case AlterTableAddColumnsStatement(tblName, cols) =>
```

Comment:
It's very weird that only `ALTER ADD COLUMN` checks temp view. This is an existing behavior so I don't want to change it in this refactor PR.